include/deal.II/meshworker/copy_data.h

```diff
@@ -0,0 +1,144 @@
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace MeshWorker
+{
+  /**
+   * Helper copy data struct.
+   *
+   * This class is a good default drop in CopyData object for the
+   * WorkStream::run() and MeshWorker::mesh_loop() functions.
+   *
+   * It allows you to store arbitrary number of (local) FullMatrices, vectors,
+   * and local_dof_indices vectors.
+   *
+   * @author Luca Heltai, 2019.
+   */
+  template <int n_matrices    = 1,
+            int n_vectors     = n_matrices,
+            int n_dof_indices = n_matrices>
+  struct CopyData
+  {
+    /**
+     * Initialize everything with the same @p size.
+     */
+    CopyData(const unsigned int size);
+
+    /**
+     * For every object, specify the size they should have.
+     */
+    CopyData(
+      const std::array<std::array<unsigned int, 2>, n_matrices> &matrix_sizes,
+      const std::array<unsigned int, n_vectors> &                vector_sizes,
+      const std::array<unsigned int, n_dof_indices> &dof_indices_sizes);
+
+    /**
+     * Deep copy constructor.
+     */
+    CopyData(const CopyData<n_matrices, n_vectors, n_dof_indices> &other) =
+      default;
+
+    /**
+     * Allow resetting of all elements of the struct to zero, by simply
+     * calling (*this) = 0;
+     */
+    void
+    operator=(const double &number);
+
+    /**
+     * Array of local matrices.
+     */
+    std::array<FullMatrix<double>, n_matrices> matrices;
+
+    /**
+     * Array of local vectors.
+     */
+    std::array<Vector<double>, n_vectors> vectors;
+
+    /**
+     * Array of local degrees of freedom indices.
+     */
+    std::array<std::vector<types::global_dof_index>, n_dof_indices>
+      local_dof_indices;
+  };
+
+
+
+  //
+  // Template definitions
+  //
+  template <int n_matrices, int n_vectors, int n_dof_indices>
+  CopyData<n_matrices, n_vectors, n_dof_indices>::CopyData(
+    const unsigned int size)
+  {
+    for (auto &m : matrices)
+      m.reinit({size, size});
+    for (auto &v : vectors)
+      v.reinit(size);
+    for (auto &d : local_dof_indices)
+      d.resize(size);
+  }
+
+
+
+  template <int n_matrices, int n_vectors, int n_dof_indices>
+  CopyData<n_matrices, n_vectors, n_dof_indices>::CopyData(
+    const std::array<std::array<unsigned int, 2>, n_matrices> &matrix_sizes,
+    const std::array<unsigned int, n_vectors> &                vector_sizes,
+    const std::array<unsigned int, n_dof_indices> &dof_indices_sizes)
+  {
+    unsigned int i = 0;
+    for (auto &m : matrices)
+      m.reinit(matrix_sizes[i++]);
+    i = 0;
+    for (auto &v : vectors)
+      v.reinit(vector_sizes[i++]);
+    i = 0;
+    for (auto &d : local_dof_indices)
+      d.resize(dof_indices_sizes[i++]);
+  }
+
+
+
+  template <int n_matrices, int n_vectors, int n_dof_indices>
+  void
+  CopyData<n_matrices, n_vectors, n_dof_indices>::
+  operator=(const double &number)
+  {
+    for (auto &m : matrices)
+      m = number;
+    for (auto &v : vectors)
+      v = number;
+  }
```

Comment:
For safety, maybe its a good idea to set the local DoF indices to an invalid number as well?